smack-websocket/src/main/java/org/jivesoftware/smack/websocket/implementations/okhttp/OkHttpWebsocketImpl.java

```diff
@@ -0,0 +1,183 @@
+ * limitations under the License.
+ */
+package org.jivesoftware.smack.websocket.implementations.okhttp;
+
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.net.ssl.SSLSession;
+
+import org.jivesoftware.smack.SmackException;
+import org.jivesoftware.smack.XMPPException;
+import org.jivesoftware.smack.c2s.internal.ModularXmppClientToServerConnectionInternal;
+import org.jivesoftware.smack.packet.TopLevelStreamElement;
+import org.jivesoftware.smack.util.PacketParserUtils;
+import org.jivesoftware.smack.websocket.WebsocketException;
+import org.jivesoftware.smack.websocket.XmppWebsocketTransportModule.XmppWebsocketTransport.DiscoveredWebsocketEndpoints;
+import org.jivesoftware.smack.websocket.elements.CloseElement;
+import org.jivesoftware.smack.websocket.elements.OpenElement;
+import org.jivesoftware.smack.websocket.implementations.AbstractWebsocket;
+import org.jivesoftware.smack.websocket.rce.WebsocketRemoteConnectionEndpoint;
+import org.jivesoftware.smack.xml.XmlPullParserException;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.WebSocket;
+import okhttp3.WebSocketListener;
+
+public final class OkHttpWebsocketImpl extends AbstractWebsocket {
+
+    private static final Logger LOGGER = Logger.getLogger(OkHttpWebsocketImpl.class.getName());
+    private static OkHttpClient okHttpClient = null;
+
+    private final ModularXmppClientToServerConnectionInternal connectionInternal;
+    private String openStreamHeader;
+    private WebSocket currentWebsocket;
+    private WebsocketConnectionPhase phase;
+    private LoggingInterceptor interceptor;
+    private WebsocketRemoteConnectionEndpoint connectedEndpoint;
+
+    public OkHttpWebsocketImpl(ModularXmppClientToServerConnectionInternal connectionInternal,
+            DiscoveredWebsocketEndpoints discoveredWebsocketEndpoints) {
+        this.connectionInternal = connectionInternal;
+
+        if (okHttpClient == null) {
+            // Creates an instance of okHttp client.
+            OkHttpClient.Builder builder = new OkHttpClient.Builder();
+            // Add some mechanism to enable and disable this intercepter.
+            if (connectionInternal.smackDebugger != null) {
+                interceptor = new LoggingInterceptor(connectionInternal.smackDebugger);
+            }
+            okHttpClient = builder.build();
+        }
+    }
+
+    @Override
+    public void connect(WebsocketRemoteConnectionEndpoint endpoint) throws InterruptedException, SmackException, XMPPException {
+        final String currentUri = endpoint.getWebsocketEndpoint().toString();
+        Request request = new Request.Builder()
+                              .url(currentUri)
+                              .header("Sec-WebSocket-Protocol", "xmpp")
+                              .build();
+
+        WebSocketListener listener = new WebSocketListener() {
+            @Override
```

Comment:
Please add newlines before `@Override` for better readability. Here and below and everywhere.